crates/scope/src/builder.rs

```diff
@@ -2380,7 +2674,36 @@ impl ScopeDataMapBuilder {
         }
     }
 
+    pub fn enter_syntax_only_mode(&mut self) {
+        assert!(self.syntax_only_depth.is_none());
+        self.syntax_only_depth = Some(self.builder_stack.depth());
+    }
+
+    fn exit_syntax_only_mode_if_necessary(&mut self) {
```

Comment:
nit: `maybe_exit_syntax_only_mode` might be better than `_if_necessary`.
The `if` condition implemented here is whether the depth matches the recorded depth at which the syntax only mode was entered.